prediction/src/preprocess/improved_lung_segmentation.py

```diff
@@ -0,0 +1,444 @@
+                                                  bronch_cts,
+                                                  levels)
+
+    seeds = np.zeros(patient.shape)
+    seeds[level] = select_seeds(bronchial, ct_slice)
+
+    return seeds
+
+
+def growing_bronchis(patient, seeds,
+                     threshold=-950,
+                     step=64,
+                     full_extraction=True):
+    """Extraction of large airways.
+
+    Before the lungs are segmented, the trachea and
+    the bronchi are extracted using the technique of a region growing.
+    To initialize the region growing, the start point is automatically
+    identified on the upper axial scan slices by searching for related
+    areas with the specified properties.
+    When an explosion occurred, the increase in threshold was divided by 2.
+
+    Args:
+        patient: a label of CT;
+        seeds: a initial region (pixel);
+        threshold: an average HU;
+        step: In each iteration, the threshold was initially increased with 64 HU
+        full_extraction: full extraction of large airways.
+
+    Returns:
+        ret - a label of the bronchi at the previous stage.
+        seeds - a label of the bronchi.
+
+    """
+
+    seeds = seeds.astype(np.bool_)
+    seeds = region_growing(patient.copy(), seeds, -5010, threshold)
+    volume = np.count_nonzero(seeds)
+
+    lungs_thresh = skimage.filters.threshold_otsu(patient[patient.shape[0] // 2])
+
+    ret = None
+    while True:
+        labeled = region_growing(patient.copy(), seeds, -5010, threshold + step)
+        new_volume = np.count_nonzero(labeled)
+        if new_volume >= volume * 2:
+            if step == 4:
+                ret = seeds.copy()
+                if not full_extraction:
+                    return ret
+
+            if step == 2:
+                return ret, seeds
+            step = np.ceil(step * 0.5)
+            continue
+
+        threshold += step
+        volume = new_volume
+        seeds = labeled
+
+        if threshold >= lungs_thresh:
+            if ret is None:
+                ret = seeds.copy()
+
+            if not full_extraction:
+                return ret
+
+            return ret, seeds
+
+
+def grow_lungs(patient, seeds):
+    """ Segmentation of lung regions.
+
+    The lungs are segmented using region growing. As a
+    seed point for the region growing, the voxel with the lowest
+    HU within the airways is used.
+
+    To determine the upper threshold for the
+    region growing operation, optimal thresholding is applied.
+
+    Args:
+        patient (np.ndarray[bool]): a label of CT
+
+    Returns:
+        np.ndarray[bool]: a label of the segmented lungs.
+
+    """
+    lungs_seeds = patient * seeds == patient[seeds].min()
+    lungs_seeds = lungs_seeds.astype(np.bool_)
+    threshold = skimage.filters.threshold_otsu(patient[patient.shape[0] // 2])
+
+    lungs_seeds = region_growing(patient.copy(), lungs_seeds, -1010, threshold)
```

Comment:
where does `-1010` come from?

Reply:
I changed this value to -1040 in a new version. This is the minimum value in the Hounsfield units that lungs can take